dangerzone/document.py

```diff
@@ -12,29 +12,57 @@ class DocumentHolder(object):
     Keeps track of one document
     """
 
-    def __init__(self, input_file_path: str) -> None:
-        # Name of input and out files
-        self._input_filename: Optional[str] = input_file_path
+    def __init__(self, input_file_path: str = None) -> None:
+        self._input_filename: Optional[str] = None
         self._output_filename: Optional[str] = None
 
+        if input_file_path:
+            self.input_filename = input_file_path
+
     @property
     def input_filename(self) -> str:
         if self._input_filename is None:
-            raise RuntimeError("Input filename has not been set yet.")
+            raise DocumentFilenameException("Input filename has not been set yet.")
         else:
             return self._input_filename
 
     @input_filename.setter
     def input_filename(self, filename: str) -> None:
+        # validate input filename
+        try:
+            open(filename, "rb")
+        except FileNotFoundError:
+            raise DocumentFilenameException(
+                "Input file not found: make sure you typed it correctly."
+            )
+        except PermissionError:
+            raise DocumentFilenameException(
+                "You don't have permission to open the input file."
+            )
+
         self._input_filename = filename
 
     @property
     def output_filename(self) -> str:
         if self._output_filename is None:
-            raise RuntimeError("Output filename has not been set yet.")
+            raise DocumentFilenameException("Output filename has not been set yet.")
         else:
             return self._output_filename
 
     @output_filename.setter
     def output_filename(self, filename: str) -> None:
+        # validate output filename
+        if not filename.endswith(".pdf"):
+            raise DocumentFilenameException("Safe PDF filename must end in '.pdf'")
+        try:
+            with open(os.path.abspath(filename), "wb"):
+                pass
+        except PermissionError:
+            raise DocumentFilenameException("Safe PDF filename is not writable")
+
```

Comment:
I fully agree with moving these checks closer to the `DocumentHolder` class. However, this way we can't reuse them from other contexts. For instance, it would make the CLI logic much simpler if we used Click validators: https://click.palletsprojects.com/en/8.1.x/options/#callbacks-for-validation. If we moved this logic into class/static methods, it would be trivial to create those.

Reply:
I tried making this a static method, but it will look weird because click's validators have three parameters (`ctx`, `param`, `value`) and not just `filename`. And we'd have to duplicate this validation code because at this moment we click for the cli and another for the gui.

Reply:
I believe this could be solved via:

* An `args.py` module that is Click-aware,
* Some `validate_{input,output}_filename()` functions, that implement the callback interface for Click (`ctx`, `param`, `value`) and wrap the `Document.validate_{input,output}_filename()` methods.
* Calling the `args.validate_{input,output}_filename()` functions from the CLI and GUI code.

Reply:
Sounds good. Yes, if you could write it, that would work for me.